sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropNamespaceExec.scala

```diff
@@ -46,9 +46,17 @@ case class DropNamespaceExec(
         }
       }
 
-      if (!nsCatalog.dropNamespace(ns)) {
-        throw QueryCompilationErrors.cannotDropNonemptyNamespaceError(namespace)
+      nsCatalog match {
+        case v2SessionCatalog: V2SessionCatalog =>
+          if (!v2SessionCatalog.dropNamespaceWithCascadeMode(ns, cascade)) {
```

Comment:
why do we need this? `dropNamespace` should drop the name in cascade mode.

Reply:
I think the problem here is how to fail quickly if the namespace is not empty and the cascade is not specified. The current implementation is quite slow because it does a `listTables`. What's worse, it does not check if the database has functions or not.

We can certainly special-case Hive here and call its `dropDatabase` method with the cascade parameter. But this exposes a limitation in the API, that the implementations can't implement `dropNamespace` well.

One option is to add a new API `isNamespaceEmpty`, so that we can simply call this API and fail immediately. However, Hive Metastore can't implement this API efficiently.

My proposal is to add a new API `dropNamespace(String[] ns, boolean cascade)` to replace the API, and ask the implementation to deal with the cascade parameter by itself.

cc @imback82 @huaxingao @MaxGekk

Reply:
> My proposal is to add a new API `dropNamespace(String[] ns, boolean cascade)` to replace the existing one, and ask the implementation to deal with the cascade parameter by itself.

+1

Reply:
+1

Reply:
Should I include the change of that API to `dropNamespace(String[] ns, boolean cascade)` in this PR? @cloud-fan

Reply:
Let's do it with a separated PR, as it's an API change.